### PR TITLE
add torch.autograd._set_view_replay_enabled, use in aot autograd

### PR DESCRIPTION
**Pull request by bdhirsh**
tldr; this should fix some minor perf regressions that were caused by adding more as_strided() calls in aot autograd.

This PR adds a new context manager, `torch.autograd._set_view_replay_enabled()`.

Context: AOT Autograd has special handling for "outputs that alias graph intermediates". E.g. given this function:

```
def f(x):
    y = torch.mul(x, 2)
    out = y.view(-1)
    return out
```

AOT Autograd will do the following:

```
def fn_to_compile(x):
    y = torch.mul(x, 2)
    out = y.view(-1)
    # return the graph intermediate
    return y, out

compiled_fn = compile(fn_to_compile)

def wrapper(x):
    y, out = compiled_fn(x)
    # regenerate the alias of the graph intermediate
    return out._view_func(y)
```

What's annoying is that `out._view_func()` will result in a `.as_strided` call, because `out` is an ordinary runtime tensor. This (likely?) caused a perf regression, because when running the backward, out `as_strided_backward()` is slower than our `view_backward()`.

In this PR, I added some TLS for instructing autograd to do view replay instead of as_strided, even when given a normal tensor. I'm definitely interested in thoughts from autograd folks (cc @albanD @soulitzer). A few points that I want to bring up:

(1) One reason that this API seems generally useful to me is because of the case where you `torch.compile()` a function, and you pass in two inputs that alias each other, and mutate one of the inputs. Autograd is forced to add a bunch of as_strided() calls into the graph when this happens, but this would give users an escape hatch for better compiled perf in this situation

(2) To be fair, AOT Autograd probably won't need this TLS in the long term. There's a better (more complicated) solution, where AOT Autograd manually precomputes the view chain off of graph intermediates during tracing, and re-applies them at runtime. This is kind of complicated though and feels lower priority to implement immediately.

(3) Given all of that I made the API private, but lmk what you all think.

This is a followup of https://github.com/pytorch/pytorch/pull/92255.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #93994
* #92857
* #94195
* #94018
* #92924
* __->__ #92588

### CODE REVIEW

**Comment by bdhirsh** [comment]
cc @ngimel. I tried benchmarking with `python benchmarks/dynamo/torchbench.py  --performance  --backend inductor --float16 --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt) --only hf_Bert`. I confirmed that we used to call `.as_strided()` in the model and now call `.view()`, although I didn't see a noticeable change in perf. Is there an easy way to test this across all of our models?

**Comment by ngimel** [comment]
Running full suites, but I think we can just wait for the dashboard to update

**Review by ezyang — APPROVED** [review]
ok

**Comment by ngimel** [comment]
Alternatively, could we improve AsStridedBackward for the cases where we can easily see that all original elements participated, so no need to prefill with 0s before copying grad?

File: torch/autograd/grad_mode.py [hunk 1]
```diff
@@ -253,3 +254,39 @@ def __exit__(self, *args) -> None:
 
     def clone(self):
         return self.__class__(self.mode)
+
+
+class _set_view_replay_enabled(_DecoratorContextManager):
```
Comment (albanD) [inline torch/autograd/grad_mode.py:259]: Could you rename this to `_force_original_view_tracking()` or something like that to be closer to what it does?

**Review by albanD — COMMENTED** [review]
Sounds ok, updating the name throughout would be nice.

**Comment by ezyang** [comment]
Yeah, AsStridedBackward improvements would also be nice

**Comment by albanD** [comment]
The 400 line comment before the function makes me afraid of diving into it https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/FunctionsManual.cpp#L2322-L2736 :D

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -30,6 +30,8 @@
 from .partitioners import default_partition
 from torch._guards import TracingContext, DuplicateInputs
 
+guard__ = torch.autograd._force_original_view_tracking(True)
```
Comment (albanD) [inline torch/_functorch/aot_autograd.py:33]: I guess this is a debug statement? This will enable original view tracking as soon as this file is imported.

File: test/test_autograd.py [hunk 3]
```diff
@@ -3717,6 +3717,21 @@ def hook(t_):
...
 
+    def test_view_replay_enabled(self):
+        def f(x):
+            out = x.clone().view(-1)
+            # mutate the view, triggering autograd view-replay logic
+            out.add_(1)
+            return out
+
+        x = torch.ones(2, 2, requires_grad=True)
+        with torch.autograd._force_original_view_tracking(True):
```
Comment (albanD) [inline test/test_autograd.py:3728]: Could you also test that without this mode, you do get an AsStridedBackward?
I feel like with the current version of this PR that would fail.

**Review by albanD — CHANGES_REQUESTED** [review]
Thanks a lot for the rename, that sounds good.
I think there is an extra guard that needs removing though

File: torch/_functorch/aot_autograd.py (hunk 2 above)
Comment (bdhirsh) [reply torch/_functorch/aot_autograd.py:33]: ha yep, this was me debugging. Thanks!

File: test/test_autograd.py (hunk 3 above)
Comment (bdhirsh) [reply test/test_autograd.py:3728]: good point, I'll add that

**Review by albanD — APPROVED** [review]
Sounds good thanks!
